### PR TITLE
[WIP] Cuda Backend

### PR DESCRIPTION
**Pull request by Steven-Roberts**
I finished merging in most of the new active/passive changes into the cuda branch, so it seems like a good time to start discussing the merge into master.  There are still a few lingering issues that will need to be resolved first.

- For efficiency, I have custom versions of `CeedBasisApply` and `CeedQFunctionApply` that accept Vectors as arguments and work over all elements at once.
- In order to be compatible with the current API, I have `CeedBasisApply` work as a light wrapper that calls my custom optimized function.  I wanted to do the same for `CeedQFunctionApply` but the current parameters don't offer enough info to call my version (missing the element strides for each input `u` and output `v`).  I figured it would be better to work on allowing batching than trying to hack this together.  Also, this is why the tests currently fail.
- Getting a cuda qfunction to the cuda backend through the c api was a bit tricky.  The solution I use uses some macros that detect if the code is being compiled with cuda.
https://github.com/CEED/libCEED/blob/f815a8afaef76789af6f4f147084698c745ec92a/include/ceed.h#L275-L307
https://github.com/CEED/libCEED/blob/6474b817cfb6eefde27f34b5603f58872ad89d9d/examples/ceed/ex1.c#L39-L40
- I haven't tried mixing fortran and the cuda backend.  There may be difficulties with getting the cuda qfunction pointer through fortran
- Perhaps these issues could be avoided by using NVRTC

### CODE REVIEW

**Comment by jedbrown** [comment]
There is one commit (0c11954d869609e040ed80e26bcdb8a753e4483c) with weird author info `= <=>`. I think it would be preferable to rebase this onto 'master' and squash into one or a very small number of commits to avoid the many broken states during development (that's fine when you're developing, but that history can't be bisected and can't be reviewed for correctness). Compare

```
$ git diff --shortstat ..origin/cuda                                                                                                                     
 14 files changed, 993 insertions(+), 45 deletions(-)                                                                                                                                         
$ git diff --shortstat ...origin/cuda                                                                                                                    
 62 files changed, 4260 insertions(+), 1282 deletions(-)        
```

This PR looks really promising, thanks!

Working over all elements at once for basis application and qfunctions is what I would expect. I think we'll actually want to be flexible about how many quadrature points a given CUDA thread works on -- I expect it will depend on number of fields, complexity of the function, and target architecture.

We should discuss the possibility of fusing BasisApply with QFunctionApply, at least for some sizes. That potentially saves two extra round trips through global memory.

Can we make the extra implementation functions `static` or `CEED_INTERN`? This sort of stuff is in `nm -D lib/libceed.so | c++filt` and either needs to be made internal/static (preferred) or namespaced if that is not possible.

```
00000000000142ba T interp(int, int, int, int, int, int, int, int, CeedTransposeMode, double const*, double const*, double*)
0000000000016754 T noTrTr(int, int, int, int, int const*, double const*, double*)
0000000000016beb T trNoTr(int, int, int, int, int, int const*, double const*, double*)
00000000000147dd T weight(int, int, double const*, double*)
```

**Comment by jedbrown** [comment]
There is an example `/opt/cuda/samples/0_Simple/simpleSeparateCompilation/simpleSeparateCompilation.cu` distributed with CUDA. I had hoped we would be able to use that approach rather than needing to use a template to fully specify the kernel at compile time. See also this [blog post](https://devblogs.nvidia.com/separate-compilation-linking-cuda-device-code/).

File: include/ceed.h [hunk 1]
```diff
@@ -215,27 +242,71 @@ CEED_EXTERN int CeedGaussQuadrature(CeedInt Q, CeedScalar *qref1d,
...
 
-static inline CeedInt CeedPowInt(CeedInt base, CeedInt power) {
+#ifdef __CUDACC__
+
+template <CeedQFunctionCallback callback>
+__global__ void apply(const CeedInt nelem, const CeedInt Q, const CeedInt numinputfields, const CeedInt numoutputfields,
+    void *ctx, const CeedScalar *const *u, CeedScalar *const *v, const CeedInt *uoffsets, const CeedInt *voffsets, int *ierr) {
+  const int i = blockIdx.x*blockDim.x + threadIdx.x;
+
+  for (int elem = i; elem < nelem; i += gridDim.x * blockDim.x) {
```
[the remark this answers is not in the document]
Comment (Steven-Roberts) [reply include/ceed.h]: Good catch.  Fixed by 69518d4349c46312f36ee2cb2591fde8be2c8630.

**Comment by Steven-Roberts** [comment]
Thanks for all the feedback @jedbrown.

- I basically did a manual merge of the active passive updates and didn't realize it would mess up the git history like that.  I redid the commit with a proper merge now.
- The very first thing I tried with cuda qfunctions was passing a device function pointer.  Based on everything I've tried, a shared object cannot call a device pointer passed to it.  I had no problem getting it to work with a static library or when it's compiled into a .o file.  I referenced https://devblogs.nvidia.com/cuda-pro-tip-use-cufft-callbacks-custom-data-processing/ to try to use device function pointers.  Note at the bottom is says the statically linked library must be used.
- Since device pointers didn't work I looked into what other libraries were doing.  I found the thrust library uses templates in these scenarios (e.g. sorting comparator).  Also, it should have better performance since it can be inlined.
- I tried fixing the commit with the weird author info.  At almost every step of the rebase I'm getting conflicts.  I'm not very familiar with rebasing, and I don't want to accidentally mess up the history.  I might need some assistance on this one.
- I distribute the quadrature points evenly to all threads for the basis kernels.  Basically each inner product in the tensor contraction is run in parallel across a thread block.  Is this what you mean by "we'll actually want to be flexible about how many quadrature points a given CUDA thread works on"
- Fusing basis and qfunction is a good optimization to try.  Any reason not to fuse the restriction as well?
- Looks like my cuda kernels were not declared static.  For some reason I thought they could have a static modifier, but they apparently can.  Fixed in 69518d4.

**Comment by YohannDudouit** [comment]
I would also be very interested to test this Cuda backend as soon as `BasisApply` and `QFunctionApply` can be fused since we already have such Cuda kernels that we implemented some time ago (using warp shuffle intrinsics).
We also attempted to fuse the scatter/gather part with `BasisApply` and `QFunctionApply` but the performance gain was marginal due to the very irregular memory access. We explored different strategies to make the scatter/gather more efficient, but we haven't found anything very satisfactory so far. Did anyone explored scatter/gather algorithm on GPU? Specialized GPU data structures or something similar?

**Comment by jedbrown** [comment]
* According to the Travis-CI reports, most intermediate commits do not pass. It's better to have one big commit than a sequence of broken commits. (It's better to have a sequence of incremental commits, all of which are tested. For example, implement the new backend for initialization, for Vector, for Basis, etc., with the tests passing at each step. But one commit is okay and normal for a new backend.) You should be able to squash everything into one commit without needing to resolve conflicts.

* I think I would rather require static libraries for the CUDA backend than to fully specify the kernel via a compile-time template. It changes the way we think about composition and I'm afraid the template approach will constrain us. (Also, it's unmaintainable if we want to fuse, especially if different backends want to fuse in different ways or a backend wants to decide how to fuse based on run-time information such as sizes `P`, `Q`, or number of elements.) In lieu of data indicating otherwise, I'm not worried about function call overhead because there should be plenty of work inside the qfunction. I understand there may be a need for qfunctions to support a strided loop.

**Comment by Steven-Roberts** [comment]
Ok I can make libCEED compile as a static library as well.  The code in tests and examples will still need to be compiled with nvcc to use the cuda backend and the code might look something like
```
CEED_SAFE_HOST_DEVICE int setup(...) { ... }

// If you know nvcc will be the compiler, you can use
__managed__ __device__ CeedQFunctionCallback setup_cuda = setup;
// but we probably should have macro to handle all cases
CEED_SAFE_MANAGED CeedQFunctionCallback setup_cuda = setup;

int main() {
  ...
  CeedQFunctionCreate(..., setup_cuda, ...);
  ...
}
```

NVRTC seems like an appealing option too, although it may be stepping on the toes of OCCA.  Any thoughts on that?

**Comment by jedbrown** [comment]
Don't worry about "stepping on toes". If NVRTC enables a better interface, we should use it.

**Comment by YohannDudouit** [comment]
I also think that we cannot really avoid NVRTC, since you need to know the order and number of quadrature point at compilation time to have decent performance on GPU. Doing a C-function callback was tricky though, as the NVRTC interface is really C++ oriented.

**Comment by jedbrown** [comment]
The qfunction does not depend on order or number of quadrature points per element, and total number of quadrature points is dynamic anyway (depends on mesh size). A backend might want to do different fusion for different sizes. We can test, but I don't think the compiler needs to know any of that to get good performance from a qfunction. I'm also skeptical that order needs to be known statically for the tensor contractions (at least within the ranges where external vectorization is used).

**Comment by YohannDudouit** [comment]
My remark was indeed more general that only applying the qfunction. If you use any array (in contractions especially), you need to know the tensor sizes at compilation time, otherwise the compiler cannot use registers 'addresses' directly in the code and you end with memory spill because the array index is a runtime information.

**Comment by jedbrown** [comment]
@YohannDudouit What code are you referring to? I'd like to make sure we're talking about the same things. It might be useful to discuss a particular snippet so we can discuss the effect of compile-time knowledge.

**Comment by YohannDudouit** [comment]
@jedbrown I'm sorry for jumping in this conversation with obscure comments, I'll try to be more specific in my next comments. For instance instance if we take this simple contraction code where the whole contraction is happening in one GPU thread: (A is the input dofs tensor, B the basis evaluated at quadrature points in 1d, and T the result)
```
inline __device__ void Contract(const real *A, const real *B,
                                 int nA1, int nA2, int nA3,
                                 int nB1, int nB2, real *T)
{
#pragma unroll
    for (int l = 0; l < nA2*nA3*nB2; l++) T[l] = 0.0;
#pragma unroll
    for (int a2 = 0; a2 < nA2; a2++)
#pragma unroll
        for (int a3 = 0; a3 < nA3; a3++)
#pragma unroll
            for (int b2 = 0; b2 < nB2; b2++)
#pragma unroll
                for (int t = 0; t < nB1; t++)
                {
                    T[a2 + a3*nA2 + b2*nA2*nA3] += B[b2*nB1 + t] * A[a3*nA2*nA1 + a2*nA1 + t];
                }
}
```
Don't pay too much attention to the codding style... This code does a contraction of a second order tensor with a third order tensor, and results in a third order tensor. We avoid here the memory spill because the number of dofs in 1d and quadrature points in 1d (size of the B tensor is `dosf1d x quads1d`) is known at compilation time, thus the array indices can be known at compilation time.
You can use the exact same code where these sizes are not known at compilation time and the performance will be terrible, since these arrays get evicted to L1, but since it's very small on GPU usually gets evicted to L2, and if we have a good occupancy then it gets evicted to Device memory. So every array access, even for temporary arrays, ends up being a global memory access which is not a viable option.

**Comment by jedbrown** [comment]
@YohannDudouit I guess you want everything to be unrolled in your example, even if it causes instruction fetch stalls due to exceeding the I-cache. If you want that, you'll need to know the sizes in your tensor contraction (which is entirely controlled by the backend, thus can be specialized with no visibility to the user), though I'm skeptical that it is desirable for large sizes. (Also, I doubt you want a single CUDA thread to do all that work.)

**Comment by jedbrown** [comment]
@stomov It would be useful for you to join the discussion about how to manage user kernels with CUDA (see above). We can create an issue and/or discuss in person this week.

**Comment by Steven-Roberts** [comment]
Im currently in the processor of switching things to NVRTC.  Seems promising thus far.

**Comment by Steven-Roberts** [comment]
The tests are still currently failing for two reasons:

- The cuda `QFunctionApply` is only implemented for vector inputs.
- Many fortran tests involving vectors fail due to issues discussed in #99.  The extra memcpy means `CeedVectorRestoreArray` doesn't recognize the pointer as an alias to either the host or device array.

**Comment by jeremylt** [comment]
I think PR #134 should help with the Fortran vector issues. If that is the case, would this put us close to bringing the CUDA backend live?

**Comment by Steven-Roberts** [comment]
@jeremylt Definitely.  I think the last issue is using vector args for the basis and qfunction.  I've mentioned it a couple of times, but I'm not sure if there is a consensus that this should be the new standard.  I've been a bit out of the loop with the work this week, so I'll need to see what needs to be merged in and updated in this branch.

**Comment by jeremylt** [comment]
I believe right now we are letting backends get that effect by passing in a pointer to the array of the vector and arguments that correspond to multiple elements.

For example:

```
ierr = CeedBasisApply(basis, nelem, CEED_NOTRANSPOSE,
              CEED_EVAL_INTERP, array_in, array_out);
```

and 

```
ierr = CeedQFunctionApply(op->qf, Q*nelem,
                          (const CeedScalar * const*) impl->indata,
                          impl->outdata); CeedChk(ierr);
```

**Comment by jedbrown** [comment]
@Steven-Roberts There is no requirement that a backend call `CeedBasisApply`. Also, since you (the CUDA backend) control the implementation of `CeedBasisApply_Cuda`, you can pass device pointers if you prefer. Maybe we should add a `CeedMemType` argument to `CeedBasisApply`?

I'm not totally sold on switching to `CeedVector` arguments because some backends iterate through small batches of elements while incrementing offsets and it's a little clumsy to wrap and unwrap that memory in tiny vectors.

**Comment by jeremylt** [comment]
I also am hesitant to use full vectors. It would involve a lot of getting and setting arrays or fields in vectors rather than directly pointing around the pointers to the memory, be that on host or device. That feels like an unneeded slowdown to me.

Adding a mem type would make sense to me , especially for a backend like OCCA that supports working on host and device. I feel like this could support a GPU keeping all memory on device for an entire OpApply. Thoughts @camierjs, @dmed256?

**Comment by jeremylt** [comment]
I think ultimately we want the same thing - letting backends apply QFuctions and Bases across multiple elements.

In order to do this, a Basis needs to know how many elements it is operating on and some way to know where the inputs and outputs are.

Action at quadrature points should be independent, so a QFunction just needs where the inputs and outputs are and a number of quadrature points to iterate over.

I think using pointers (which may be on device or host) and a number of elements/quadrature points allows that right now. So the question is if wrapping those pointers in a vector is a better interface. We have more CPU backends coming, so this interface needs serve both CPU and GPU backends.

I don't understand what wrapping these pointers in vectors buys us. I'm not seeing what it enables that we can't currently do. Backends already have the option to put and keep data on the device for as long as they want, or at least that is the intention with the current interface.

Like I said, I think we want the same thing - multiple element operations and reduced memory motion. It just seems like it's possible under the current interface and it seems like switching to vector inputs and outputs doesn't add capability but does add complexity. Or am I missing something?

**Comment by v-dobrev** [comment]
Here is my opinion:

In the user interface, we should only support `CeedBasisApply` that works over all elements - it maps between E-vectors and Q-vectors.

Internally, each backend can define and use its own versions of the function - e.g. a version that works on batches of elements - if that makes sense for the backend. The user interface should not require such low-level functions to be exposed, because they do not make sense (from efficiency perspective) with some backends (e.g. GPU backends).

This argument, I think, also applies to the function `CeedQFunctionApply`.

**Comment by jeremylt** [comment]
I am not sure we intend users to use BasisApply or QFunctionApply in a typical situation. We perhaps should have a discussion about what we consider 'backend' functions v.s. 'user' functions. We started this sort of discussion in the Doxygen by tagging some functions as intended for developers.

I think a backend API needs to expose a number of elements to operate over (to include all of them) to flexibly support a wider range of backends, but I agree that a user should not need to know about that.

**Comment by Steven-Roberts** [comment]
@jeremylt Yes, I think we have the same goals here. It seems the issue of working over multiple elements has been resolved.  Let me summarize the benefits of using `CeedVector`

- If we stick to host or device pointers the caller will need to know which one is used.  If `BasisApply` and `QFunctionApply` is kept public, it seems like users shouldn't need to know that low-level detail.
- OCCA doesn't represent the device array with `CeedScalar*`
- Direct calls to `BasisApply` and `QFunctionApply` are faster for GPU backends (if kept public)
- The API is more consistent.  Or maybe the elem restriction should accept pointers.
- Many operator backends like ref and cuda will no longer need to read and restore vectors.

**Comment by jeremylt** [comment]
I suppose the heart of the question then is twofold:
- What is the best way to represent 'where' data is for BasisiApply and QFunctionApply
-- Perhaps a pointer isn't the easiest way to communicate 'where' data live on a device?
-- I was under the impression that we were supporting letting GPU backends keep data on the device throughout OpApply. You don't seem to think that is possible right now.
- What do we intend a 'typical' user to call
-- I don't see any of the 'Apply' functions except OpApply as being called by 'typical' users
-- I think there is benefit in allowing advanced users to access functions that backends typically use
-- I think we want our backend facing functions to be as flexible as possible, not implying a 'standard' way of handling BasisApply (such as acting on the vector of all inputs)

**Comment by jedbrown** [comment]
Users don't normally call `CeedBasisApply` so in one sense we could remove it from the interface entirely. On the other hand, we would like for developers of solvers (for example) to be able to deconstruct operations at finer granularity if need be (e.g., to obtain eigen-estimates for single/few-element subdomains with various boundary conditions). I expect this sort of development will take place first for CPU environments, though we don't want to cripple other environments. Note that patterns can graduate to the library if they benefit -- the developer cost is that each backend would be expected to implement that operation with good performance.

We want to be able to apply QFunctions and Basis at fine granularity. I think the current arguments help convey that it is a lower level interface. If we promote it to working on Vectors, we'll be obligated to operate on a range of layouts and I think that is needless complexity at this time.

Backends really shouldn't feel bad about calling their internal functions directly. The public interfaces are meant to provide a way to isolate operations for research purposes and to perform some common error checking. I don't think there will be a one-size-fits-all interface for those finer granularity operations as called within performance-sensitive code. A GPU backend will want to fuse and coalesce/overlap lots of operations that are currently separate, so the needs we are discussing in this thread sound more like conveniences for a proof-of-concept that will go away for an optimized GPU backend. If you think this is not true, i.e., you believe that a near-optimal GPU backend can implement `CeedOperatorApply` calling exclusively through the public interfaces, then we can reassess.  If not, let's just let the public interface be for now and call your own internal interfaces.

I'd really like to gain experience with an optimized GPU backend before making changes to the public interface. We need a way to experiment easily and doing that in the public interface is much more expensive and disruptive. So let's prototype whatever you want most via internal interfaces for your backend, then we can evaluate them and consider changing or augmenting the public interfaces. How does that sound?

**Comment by Steven-Roberts** [comment]
@jedbrown I understand wanting to delay the decision a bit.  I'm not sure what you mean by "If we promote it to working on Vectors, we'll be obligated to operate on a range of layouts and I think that is needless complexity at this time".  Why would the layout be any different?  This doesn't appear to be a problem for elem restriction.

**Comment by jedbrown** [comment]
@Steven-Roberts The vectorized CPU backend reorders and pads the element restriction so its "E-vector" has a complicated ordering if you wanted to index it generically (it just happens to be good for vectorization within each batch). Also, it will soon not have a complete E-vector at any time (just a workspace that fits in cache). It won't call `CeedElemRestrictionApply` because that interface has too coarse granularity for the cache locality. I don't want the public interface to imply that intermediate storage should exist in any particular representation within a backend, but I do want some finer-grained operations available for research.

It actually wouldn't bother me if some backends did not implement `CeedElemRestrictionApply`, `CeedBasisApply`, or `CeedQFunctionApply` so long as their `CeedOperatorApply` worked well. But those are still useful for unit testing (if a backend implements it, the unit tests are already written; this saves time debugging) and for research.

**Comment by YohannDudouit** [comment]
@Steven-Roberts I think that what Jed means is that ultimately an optimized Cuda backend can only be at `CeedOperator` level, you just give a pointer to the GPU Kernel (CeedOperator_private->Apply I guess) and this is the only CPU call that you do.
@jedbrown Is it compatible with a strategy where everything is stored at E-Vector level on GPU, computation is done on GPU, but the Scatter/Gather (that you call Restriction I think) is done in parallel on CPU while the GPU is computing?

**Comment by v-dobrev** [comment]
> It actually wouldn't bother me if some backends did not implement `CeedElemRestrictionApply`, `CeedBasisApply`, or `CeedQFunctionApply` so long as their `CeedOperatorApply` worked well.

@jedbrown, at this time, if a backend chooses not to implement the first tree functions, is it going to pass the tests?

**Comment by jeremylt** [comment]
Right now, the only tests a backend can skip are the ones related to batched restrictions. Eventually I plan on making a whitelist of sorts where a developer can list what tests should be skipped for a particular backend.

**Comment by v-dobrev** [comment]
> We can easily add blacklist and/or whitelist features to the test suite. Is that something you need/want now?

If this behavior is allowed e.g. by allowing `basis->Apply` to be `NULL` here:
https://github.com/CEED/libCEED/blob/0948660574688e7f772302e3ec183ecf15d81fe4/interface/ceed-basis.c#L513-L515
then this PR can be merged without the need to implement that function.

Isn't this (the debate about `CeedBasisApply`) one of the reasons for delaying the merge of this PR? That, and the exact same issue with `QFunctionApply`?

**Comment by jedbrown** [comment]
1. We can't merge this PR as-is because compilation of the test suite fails even though the CUDA backend is not used.

2. We can't just make `CeedBasisApply` "succeed" when not implemented because the interface does not know generically how the backend handles vectors. We also can't just do nothing and report success.

3. We could blacklist the CUDA backend so that Basis and QFunction tests are skipped.

4. If `CeedOperatorApply` was made to work (remember that the tests don't compile in the current PR; I don't know the status of work that hasn't been pushed yet) in the tests and examples, I would be okay merging it with a blacklist for the Basis and QFunction tests. Note that this lack of unit tests would be a calculated maintenance burden for any developer who wishes to extend the backend or to make interface changes that would require maintenance of the backend -- this work is more expensive if we only have integration tests. CUDA-specific unit tests would help mitigate this.

**Comment by Steven-Roberts** [comment]
Once `CeedQFunctionApply` is implemented all the tests could pass (assuming #134 takes care of the other issue) and this could be merged.  After reviewing the occa version, it seems possible to make `CeedQFunctionApply` convert the arrays to vectors and call `CeedQFunctionApply_Cuda`.  Unfortunately, the performance will significantly worse than what it once was, but eventually fusing will take care of that.  I start school this week so it might be a while before I can work on that.

**Comment by jedbrown** [comment]
Okay, that sounds good. I wouldn't worry about performance for any of the operations except `CeedOperatorApply`. Perhaps one of your colleagues at LLNL can help contribute to wrapping up this PR if you're pressed for time now.

**Comment by YohannDudouit** [comment]
@jedbrown We explored different strategies for the Scatter/Gather, but the inherent nature of this operation makes it very unsuited for GPUs (memory accesses are very random). NVLink2 is still 300GB/s, so about a third of the Device memory bandwidth (on Volta), thus if the shared degrees of freedom are less than a third of the total number of degrees of freedom, we can optimistically hope to completely hide the scatter/gather. In our current kernels the scatter/gather is more than half of the computation time, hiding the scatter/gather would thus be very beneficial. This also allows to keep all the connectivity information on host memory, and reuse your batched scatter/gather. I agree that this approach might not be interesting for low order. If you want to share an approach that sound more promising to you...
PS: If the data size is small, GPU are not interesting anyway, GPUs need to be "saturated" to be interesting, performance drop is huge otherwise.

**Comment by jedbrown** [comment]
@YohannDudouit If scatter/gather (ElemRestriction) is more than half your time currently, then there isn't much to hide it behind, considering that the inner element processing can't start until restriction has been applied to those elements. Perhaps it makes sense if you have enough data to fill device memory. Note that 300 GB/s is the sum of theoretical in+out.

**Comment by YohannDudouit** [comment]
@jedbrown The restriction is more than half of the computation time because it is highly inefficient on GPU, I don't have in mind the efficiency of the global load/store for the restriction, but it's bad, basically when you order a cacheline you use a small portion of it, I think that this wouldn't be as much of a problem on CPU.

**Comment by jedbrown** [comment]
If your L-vector is poorly ordered you'll have the same problem on a CPU. STREAM in strided ordering is at least 8x slower and worse in random ordering because there isn't an easy alternative way to cover DRAM latency. Experiments would be interesting.

**Comment by jeremylt** [comment]
This PR has been updated in PR #175

**Comment by Steven-Roberts** [comment]
Thanks for wrapping up this PR.  I wish I could have been more helpful, but I've been quite busy since the internship ended.

**Comment by jeremylt** [comment]
Totally understandable. @YohannDudouit doing all of the hard work lining up your backend with some interface changes, so he deserves the credit.

**Pull request closed without merging**